designs/0030-complex-numbers.md

```diff
@@ -0,0 +1,355 @@
+
+### CmdStan output
+
+CmdStan output just mirrors the C++ class output and will write headers the same way they are produced by the C++ class.
+
+### JSON input
+
+The key design aspect here is that we know when we want to read an array and know when we want to read complex numbers.  So we do not need to encode the complex/non-complex distinction directly in the output structure.  Specifically, we will follow the example in the 
+[Python JSON spec](https://docs.python.org/3/library/json.html), which encodes a complex number `3 - 1.5j` as the list `[3, -1.5]`.  
+
+A self-documenting JSON format might look like this, 
+
+```
+{ "real": 3, "imag": -1.5 }
+```
+
+but that's going to be far too verbose.
+
+### R dump format input
+
+We can adopt the standard R dump format for complex number output, which separates the real and imaginary component with the 
+
+```
+z1 <- 3-1.5i
+z2 <- 2.3+4.7i
+```
+
+## Rationale and alternatives
+
+Most of these proposals are just applying the simplest possible thing that is mathematically and computationally coherent.  For example, it's natural to promote `real` to `complex` in mathematics, so we allow it in the language.  Similarly, it's natural to 
+
+### Container type alternative
+
+It might be tempting to try to decompose types as C++ does, for example using `matrix<real>` and `matrix<complex>` or even `complex matrix`, etc.  One obstacle to this is that we use `<...>` for constraints, as in `matrix<lower = 0>`.  And we want to start stacking constraints, as in `real<multiplier = 7, offset = 3.9><lower = 0>`.  But the real reason not to do this is that there's not much shared between the matrices other than constructors, getters, and setters.
+
+For example, complex numbers do *not* work with constraints.  It does not make sense to write `complex<lower = 0> z` or `complex<offset = 7>`.  
+
+### Full covariance
+
+The biggest issue is how much covariance do we want to support given that it's not yet fully supported for `int` and `real`.  For example, it'd be nice to allow assignment of `int[]` to `real[]` and of `real[]` to `complex[]`.
+
+### Constructor alternatives
+
+We could use a simple pair structure instead of `complex(r, i)`, such as `{ r, i }`, `<r, i>`, or `[r, i]`, e.g.,
+```
+complex z = { r, i };
+```
+but these notations are already used for arrays, constraints, and row vectors respectively.
+
+
+### Accessor alternatives 
+
+For accessors, we could use method-like notation, such as `z.real()`, but Stan does not use this kind of object-oriented notation elsewhere and it is less clear that the result can be used an lvalue.
+
+### Offset and multiplier addition
+
+In addition to lower and upper bound constraints, it'd be possible to use complex offsets and multipliers, 
+```
+complex<offset = o, multiplier = m> z;
+```
+Here, the offset and multiplier could be complex scalars or real scalars or integers.
```

Comment:
Also not implemented, and in an above paragraph it is stated this would not make sense

Reply:
Right.  We should just remove it.